### ISSUE TITLE
SSLHandshakeException with system DownloadManager

### ISSUE THREAD

**Issue by daimoniuma** [skyguy126/Xposed-SoundCloudDownloader#5]
I tried enough times, and your app is always failing.
I used Download Redirection module to get link created by your module in ADM Pro:
https://api-mobile.soundcloud.com/tracks/soundcloud:tracks:248754416/streams/https?client_id=DELETED&oauth_token=DELETED

Then I tried One Tap Video Download which works well with SoundCloud:
https://api.soundcloud.com/tracks/248754416/stream?client_id=DELETED&oauth_token=DELETED&can_snip=false

Did you really tested it?

**Comment by skyguy126** [comment]
Yes it works fine for me. Maybe the downloads are blocked in your region? Not too sure what the issue is.

**Comment by daimoniuma** [comment]
I tried it again. Can't connect via system download manager. ADM can't resolve it too. Seems like api-mobile is blocked for me, if it is working for you. Please, could you add in prefs of the module api.sc.com which is available for me? Didn't know that SoundCloud provides 2 APIs. I remember that there is API v2. Maybe, your API is obsolete v1...

**Comment by skyguy126** [comment]
I'm not making any api call myself. I'm simply sending the URL used to stream the song in the official app to the Android download manager instead. Is your SoundCloud app the latest version, any other modules interfering with this one?

If all else fails I will look into adding a modified URL option.

**Comment by daimoniuma** [comment]
Sorry, but really "/streams/https?" and not "/stream?"? Just strange link. 😑

**Comment by daimoniuma** [comment]
I use One Tap. I told you. It have hook for SoundCloud. But it just link listener. It shouldn't change anything important. I'll try to disable it. Of course, I'm on latest SC.

**Comment by skyguy126** [comment]
This module is also just a link listener. Maybe I can send you a debug version to try later?

**Comment by daimoniuma** [comment]
No. One Tap was disabled. Just can't connect to your links. 😔 Anyway you see link. Client ID & Token are identical. IDK. Of course you can send debug to me.

**Comment by daimoniuma** [comment]
I tried VPN (USA). Same result. In the logcat I found this:
```
07-14 14:45:38.003 W/DownloadManager(17298): Aborting request for download 1461: while trying to execute request: javax.net.ssl.SSLHandshakeException: Handshake failed
07-14 14:23:59.752 W/DownloadManager(24166): com.android.providers.downloads.StopRequestException: while trying to execute request: javax.net.ssl.SSLHandshakeException: Handshake failed
```

**Comment by daimoniuma** [comment]
I tried this links on the notebook. Firefox does it well. Failure in the DownloadManager configuration.

**Comment by skyguy126** [comment]
I'll try and add an option to open the download link in browser rather than using download manager (although the browser will inevitably refer back to the download manager but might fix the issue).

**Comment by daimoniuma** [comment]
I'm not sure how it works on pocket Firefox, but laptop Firefox didn't suggest downloading. It just opened audio streaming in his own player.

Also you didn't escape forbidden symbols for file names. For example, 8Dio Productions uses ":" too much.

I successfully solved all issues by integrating your code into my Xposed module which have it's own download service (OkHttp). I also changed Xposed part a bit to show files size inline download option.

Thank you for initial investigation. 😉

**Comment by skyguy126** [comment]
Is your modified code on GitHub, would love to see what you changed to fix it and how I can incorporate it into the standalone Xposed build.

**Comment by daimoniuma** [comment]
I didn't modify your code in fact. I just got entry-points from your code, then it was written on my own. I also hooked other methods for "setupMenu", because I introduced option for size showing inline download button but "track" isn't loaded at "setupMenu", for example...
```
public static String getValidatedFilename(String filename) {
        StringBuilder filenameBuilder = new StringBuilder(filename);
        for(int i = 0; i < filename.length(); i++) {
            char j = filename.charAt(i);
            String reservedChars = "?:\"*|/\\<>";
            if(reservedChars.indexOf(j) != -1) {
                filenameBuilder.setCharAt(i, ' ');
            }
        }
        return filenameBuilder.toString().trim();
    }
```
It can help you with https://soundcloud.com/8dawn/8dio-free-you-cloud-collaborationblack-sands-by-lex-dumitru

Download Service of my app is much bigger than whole your app and closed sourced at the end, sorry. 😛

**Comment by skyguy126** [comment]
Ok, think I figured out the issue with https you were originally having (finally got a build environment setup):

`String url = (String) XposedHelpers.callMethod(urlBuilder, "buildHttpsStreamUrl", new Class[]{XposedHelpers.findClass("com.soundcloud.android.model.Urn", lpparam.classLoader)}, urn);`

That is the line of code that retrieves the stream url for downloading purposes. Here we specifically call **buildHttpsStreamUrl** which ensures that the stream is sent over an encrypted connection to prevent tampering, hence why you mentioned earlier, *"Sorry, but really "/streams/https?" and not "/stream?"? Just strange link."*

Currently, I do not have access to the decompiled source of the apk but I can assume there are other methods that can create a *non-https url* so to say. As of now unfortunately I don't see a reason to switch the url scheme unless I get more reports of the same error.

Added a filename validation method, thanks for the pointer.
